modules/src/ics07_tendermint/client_def.rs

```diff
@@ -25,17 +47,84 @@ impl ClientDef for TendermintClient {
+        let header_consensus_state = ConsensusState::from(header.clone());
+        let existing_consensus_state =
+            match maybe_read_consensus_state(ctx, &client_id, header.height())? {
+                Some(cs) => {
+                    // If this consensus state matches, skip verification
+                    // (optimization)
+                    if cs == header_consensus_state {
+                        // Header is already installed and matches the incoming
+                        // header (already verified)
+                        return Ok((client_state, cs));
+                    }
+                    Some(cs)
+                }
+                None => None,
+            };
+
+        let trusted_consensus_state = read_consensus_state(ctx, &client_id, header.trusted_height)?;
+
+        let trusted_state = TrustedBlockState {
+            header_time: trusted_consensus_state.timestamp,
+            height: header
+                .trusted_height
+                .revision_height
+                .try_into()
+                .map_err(|_| {
+                    Ics02Error::tendermint_handler_error(Error::invalid_header_height(
+                        header.trusted_height,
+                    ))
+                })?,
+            next_validators: &header.trusted_validator_set,
+            next_validators_hash: trusted_consensus_state.next_validators_hash,
+        };
+
+        let untrusted_state = UntrustedBlockState {
+            signed_header: &header.signed_header,
+            validators: &header.validator_set,
+            // NB: This will skip the
+            // VerificationPredicates::next_validators_match check for the
+            // untrusted state.
+            next_validators: None,
+        };
+
+        let options = client_state.light_client_options()?;
+
+        match self
+            .verifier
+            .verify(untrusted_state, trusted_state, &options, Time::now())
+        {
+            Verdict::Success => {}
+            Verdict::NotEnoughTrust(voting_power_tally) => {
+                return Err(Error::not_enough_trusted_vals_signed(format!(
+                    "voting power tally: {}",
+                    voting_power_tally
+                ))
+                .into())
+            }
+            Verdict::Invalid(detail) => {
+                return Err(Ics02Error::tendermint_handler_error(
+                    Error::verification_error(detail),
+                ))
+            }
         }
 
-        // TODO: Additional verifications should be implemented here.
+        // If the header has verified, but its corresponding consensus state
+        // differs from the existing consensus state for that height, freeze the
+        // client and return the installed consensus state.
+        if let Some(cs) = existing_consensus_state {
+            if cs != header_consensus_state {
+                return Ok((client_state.with_set_frozen(header.height()), cs));
+            }
+        }
```

Comment:
I think we are still missing here two time monotonicity checks for update-in-the-middle case (to keep things simple, these would/ should not be part of the tendermint light client check above)
Let's say we had these consensus states before the update client came:
`{.., cs-trusted, cs-lower, cs-higher, cs-latest}`
Assume that after we process the header and create a new state we would have:
`{.., cs-trusted, cs-lower, cs_new, cs-higher, cs-latest}`
We need to check:
`cs-lower.time < cs_new.time < cs-higher.time`

Reply:
This should be taken care of now by https://github.com/informalsystems/ibc-rs/pull/1321/commits/1a0a7aa61542c4de0c2b12b4a3b13b752c74b502